components/clm/src/main/initGridCellsMod.F90

```diff
@@ -123,78 +126,110 @@ subroutine initGridcells
        ci = bounds_clump%begc-1
        pi = bounds_clump%begp-1
        
-       ! For each gridcell in clump, create the correct number of topounits
-       ! As a preliminary implementation, every gridcell has the same number of topounits,
-       ! and each topounit on the gridcell has an equal weight.
+       ! For each gridcell in clump, create the correct number of topounits       
        do gdc = bounds_clump%begg, bounds_clump%endg
-          do topounit = 1, max_topounits
-             wttopounit2gridcell = 1._r8/(max_topounits)
-             call add_topounit(ti=ti, gi=gdc, wtgcell=wttopounit2gridcell)
+          tmp_tpu = ldomain%num_tunits_per_grd(gdc)       ! Actual number of topounits per grid
+          tmp_tpu2 = num_tunit_per_grd(gdc)
+          tmp_msk = ldomain%mask(gdc)
+          if(has_topounit .and. max_topounits > 1) then
+             ntopos = tmp_tpu                                
+          else 
+             ntopos = max_topounits
+          endif
+          
+          do topounit = 1, ntopos                    ! use actual/valid # of topounits per grid intead of max_topounits
+             if (max_topounits == 1) then
+                 wttopounit2gridcell = 1.0           ! The weight of topounit is 1 if only 1 topounit per grid
+                 is_tpu_active = .true.              ! Make topounit active if only one topounit is in a grid
+             else
+                 wttopounit2gridcell = wt_tunit(gdc,topounit) !grc_pp%tfrc_area(gdc,topounit) 
+                 !if (topounit <= num_topo_tmp) then
+                 if (wttopounit2gridcell > 0.0) then
+                     is_tpu_active = .true.
+                 else
+                     is_tpu_active = .false.
+                 endif                    
+             endif
+             elv = elv_tunit(gdc,topounit) !grc_pp%televation(gdc,topounit) 
+             slp = slp_tunit(gdc,topounit) !grc_pp%tslope(gdc,topounit) 
+             asp = asp_tunit(gdc,topounit) !grc_pp%taspect(gdc,topounit) 
+             topo_ind = topounit
+             !wttopounit2gridcell = 1._r8/(max_topounits)
```

Comment:
Please remove commented out lines

Reply:
@bishtgautam thanks! Commented out code has been removed.

Reply:
Thanks @bishtgautam ! I have addressed all your comments in the latest commit.